fuzz_test.go

```diff
@@ -45,112 +46,130 @@ func TestFuzz_basic(t *testing.T) {
+		s.Stage(2, obj.I16 != 0)
+		s.Stage(3, obj.I32 != 0)
+		s.Stage(4, obj.I64 != 0)
+		s.Stage(5, obj.U != 0)
+		s.Stage(6, obj.U8 != 0)
+		s.Stage(7, obj.U16 != 0)
+		s.Stage(8, obj.U32 != 0)
+		s.Stage(9, obj.U64 != 0)
+		s.Stage(10, obj.Uptr != 0)
+		s.Stage(11, obj.S != "")
+		s.Stage(12, obj.B == true)
+		s.Stage(13, !obj.T.IsZero())
+		s.Stage(14, obj.C64 != 0)
+		s.Stage(15, obj.C128 != 0)
+		return s
+	})
 }
 
 func TestFuzz_structptr(t *testing.T) {
-	obj := &struct {
+	obj := struct {
 		A *struct {
 			S string
 		}
 	}{}
 
 	f := New().NilChance(.5)
-	failed := map[string]int{}
-	for i := 0; i < 20; i++ {
-		f.Fuzz(obj)
-
-		if n, v := "a not nil", obj.A; v == nil {
-			failed[n] = failed[n] + 1
-		}
-		if n, v := "a nil", obj.A; v != nil {
-			failed[n] = failed[n] + 1
+	tryFuzz(t, f, &obj, func() Stages {
+		s := DeclareStages(3)
+		s.Stage(0, obj.A == nil)
+		if s.Stage(1, obj.A != nil) {
+			s.Stage(2, obj.A.S != "")
 		}
-		if n, v := "as", obj.A; v == nil || v.S == "" {
-			failed[n] = failed[n] + 1
-		}
-	}
-	checkFailed(t, failed)
+		return s
+	})
 }
 
-// tryFuzz tries fuzzing up to 30 times. Fail if check() never passes, report the highest
-// stage it ever got to.
-func tryFuzz(t *testing.T, f fuzzer, obj interface{}, check func() (stage int, passed bool)) {
+// tryFuzz tries fuzzing until check returns passed == true, or up to 100
+// times. Fail if check() never passes, report the highest stage it ever got
+// to.
+func tryFuzz(t *testing.T, f fuzzer, obj interface{}, check func() Stages) {
 	t.Helper()
-	maxStage := 0
-	for i := 0; i < 30; i++ {
+	var s Stages
+	for i := 0; i < 100; i++ {
 		f.Fuzz(obj)
-		stage, passed := check()
-		if stage > maxStage {
-			maxStage = stage
-		}
-		if passed {
+		got := check()
+		s = s.OrPasses(got)
+		if s.AllPassed() {
 			return
 		}
 	}
-	t.Errorf("Only ever got to stage %v", maxStage)
+	t.Errorf("Not all stages passed:\n%s", s)
+}
+
+// Stages tracks pass/fail for up to 63 stages
+type Stages struct {
+	mask   uint64
+	passed uint64
+	failed uint64
+}
+
+func DeclareStages(N uint) Stages {
+	if N >= 63 {
+		panic("the math below only works up to 63")
+	}
+	return Stages{
+		// Exercise for the reader: make this work for all 64 bits.
+		mask: (uint64(1) << N) - 1,
+	}
+}
+
+func (s Stages) String() string {
+	explain := map[uint64]string{
+		0: "never passed",
+		1: "passed",
+	}
+	var parts []string
+	for u := uint(0); s.mask != 0; u++ {
+		parts = append(parts, fmt.Sprintf("stage %v: %v", u, explain[s.passed&1]))
+		s.mask >>= 1
+		s.passed >>= 1
+	}
+	return strings.Join(parts, "\n")
+}
+
+func (s Stages) OrPasses(s2 Stages) Stages {
+	if s.mask == 0 {
+		// Allow accumulating without knowing the number of stages in
+		// advance
+		s.mask = s2.mask
+	}
+	if s.mask != s2.mask {
+		panic("coding error, stages are differently typed")
+	}
+	s.passed |= (s2.passed & ^s2.failed)
+	return s
+}
+
+func (s Stages) AllPassed() bool {
+	return (s.passed & ^s.failed) == s.mask
```

Comment:
Do you need to track failed if you have the mask? Wouldn't `s.passed == s.mask`  tell you if every stage passed at least once?  Maybe I'm not fully understanding all of these binary operations.

Reply:
Yeah, some of the code I replaced failed if any loop iteration failed. So it requires at least one pass and no failures to reproduce the prior logic. The tests could probably be adjusted to make this simpler, but that can happen later :)